j1939/Dm14Server.py

```diff
@@ -0,0 +1,283 @@
+from enum import Enum
+import queue
+import secrets
+import j1939
+
+
+class ResponseState(Enum):
+    IDLE = 1
+    WAIT_FOR_DM14 = 2
+    WAIT_FOR_KEY = 3
+    SEND_PROCEED = 4
+    SEND_OPERATION_COMPLETE = 5
+    WAIT_OPERATION_COMPLETE = 6
+    SEND_ERROR = 7
+    WAIT_FOR_DM16 = 8
+
+
+class DM14Server:
+    def __init__(self, ca: j1939.ControllerApplication) -> None:
+        """
+        performs memory access responses using DM14-DM18 messaging.
+        :param obj ca: j1939 controller application
+        """
+
+        self._ca = ca
+        self._busy = False
+        self.sa = None
+        self.state = ResponseState.IDLE
+        self._key_from_seed = None
+        self.data_queue = queue.Queue()
+        self.mem_data = None
+        self._seed_generator = self.generate_seed
+        self.address = None
+        self.length = 8
+        self.proceed = False
+        self.data = []
+        self.error = 0x00
+        self.edcp = 0x07
+        self.status = j1939.Dm15Status.PROCEED.value
+        self.direct = 0
+
+    def _wait_for_data(self) -> None:
+        """
+        Determines whether to send data or wait to receive data based on the command type.
+        If the command is a read command, then the data requested is sent.
+        """
+        if self.command is j1939.Command.READ.value:
+            self._send_dm15()
+            self._send_dm16()
+            self.proceed = True
+            self.state = ResponseState.SEND_OPERATION_COMPLETE
+            self._ca.subscribe(self.parse_dm14)
+            self._send_dm15()
+        else:
+            self._ca.subscribe(self._parse_dm16)
+            self._send_dm15()
+            self.state = ResponseState.WAIT_FOR_DM16
+
+    def parse_dm14(
+        self, priority: int, pgn: int, sa: int, timestamp: int, data: bytearray
+    ) -> None:
+        """
+        parse DM14 message received
+        :param int priority: priority of the message
+        :param int pgn: parameter group number of the message
+        :param int sa: source address of the message
+        :param int timestamp: timestamp of the message
+        :param bytearray data: data of the PDU
+        """
+        if pgn != j1939.ParameterGroupNumber.PGN.DM14:
+            return
+        if (
+            (self.sa is not None and sa != self.sa)
+            or (
+                self.address is not None and self.address != data[2 : (self.length - 2)]
+            )
+            or self._busy
+        ):
+            sa_temp = self.sa
```

Comment:
Having to store all these variables in temp variables seems a little hacky.  Maybe you should just make `_send_dm15()` take args instead?